### ISSUE TITLE
[1.7.x] VirtualService HTTP routes are not picked up when Gateway has httpsRedirect set to true

### ISSUE THREAD

**Issue by andrascz** [istio/istio#27315]
**Bug description**
When upgrading istio from version 1.6.3 to 1.7.x (tried both 1.7.0 and 1.7.1), all our virtual services gone ineffective. The ingress gateway responds with 404 to all requests.

I assume this is broken by https://github.com/istio/istio/pull/24958
A similar issue is already reported, although with a different usecase: https://github.com/istio/istio/issues/27157

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
The ingress gateway routing traffic to the backend services based on the Virtual Service routes.

**Steps to reproduce the bug**
Our AWS load balancer terminates TLS. It routes traffic to our ingress gateway service both on port 80 and 443 while also sets the appropriate `X-Forwarded-Proto` header.

Our gateway is the following:
```yaml
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: production-gateway
spec:
  selector:
    istio: ingressgateway # use istio default controller
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
    hosts:
    - "HOST.EXAMPLE.COM"
```
We have multiple virtual services all matching the same host but with different prefixes:
```yaml
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: admin-virtual-service
spec:
  hosts:
    - "HOST.EXAMPLE.COM"
  gateways:
    - production-gateway
  http:
    - match:
      - uri:
          prefix: "/admin/api/"
      route:
        - destination:
            host: admin-api-service
            port:
              number: 8080
      headers:
        ...
    - match:
      - uri:
          prefix: "/admin"
      route:
        - destination:
            host: admin-ui-service
            port:
              number: 80
      headers:
        ...
```
removing `tls.httpsRedirect: true` from the gateway configuration makes the virtual services work again.

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
```shell
➜ ./istioctl version --remote
client version: 1.7.1
control plane version: 1.7.1
data plane version: 1.7.1 (23 proxies)

➜ kubectl version --short 
Client Version: v1.18.5
Server Version: v1.17.9

➜ helm version
version.BuildInfo{Version:"v3.2.4", GitCommit:"0ad800ef43d3b826f31a5ad8dfbb4fe05d143688", GitTreeState:"clean", GoVersion:"go1.14.4"}
```

**How was Istio installed?**
Generated manifests then installed via helmfile as chart.
Config:
```yaml
apiVersion: install.istio.io/v1beta1
kind: IstioOperator
spec:
  addonComponents:
    prometheus:
      enabled: false
  profile: default
  tag: 1.7.1-distroless
  values:
    global:
      jwtPolicy: first-party-jwt
      tracer:
        zipkin:
          address: jaeger-collector.istio-system:9411
    gateways:
      istio-ingressgateway:
        runAsRoot: true
  components:
    ingressGateways:
    - enabled: true
      name: istio-ingressgateway
      k8s:
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
          service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "60"
          service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "True"
          service.beta.kubernetes.io/aws-load-balancer-internal: "true"
          service.beta.kubernetes.io/aws-load-balancer-ssl-cert: <redacted>
          service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: <redacted>
          service.beta.kubernetes.io/aws-load-balancer-ssl-ports: https
          service.beta.kubernetes.io/aws-load-balancer-type: elb

        service:
          ports:
          - name: http2
            port: 80
            targetPort: 80
          - name: https
            port: 443
            targetPort: 80
```

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS

**Comment by esnible** [comment]
Duplicate of https://github.com/istio/istio/issues/27157 ?

**Comment by andrascz** [comment]
Because in #27157 the proposed solution is to write an analyzer, I am afraid it will not solve our issue. We would like to have both HTTP routing and HTTP to HTTPS redirection on the same gateway port with HTTP protocol. The redirection happens when the `X-Forwarded-Proto` header contains `HTTP`, the routing happens when it contains `HTTPS`

**Comment by howardjohn** [comment]
Right this seems like a related but more severe issue. The other one was a misconfiguration. This seems a proper configuration by the user but invalid config sent to Envoy.

**Comment by gertvdijk** [comment]
I'm affected too, this triggered for me after upgrade 1.6.7 / 1.6.12 -> 1.7.3, on a 1.18.10 Kubernetes kubeadm self-installed environments.

Plaintext HTTP isn't redirected anymore. Happy to help out with testing a potential fix or to provide more details if needed.

**Comment by zufardhiyaulhaq** [comment]
@andrascz I believe this is expected behavior. since you don't define port 443 in the gateway itself?
```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: production-gateway
spec:
  selector:
    istio: ingressgateway # use istio default controller
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
    hosts:
    - "HOST.EXAMPLE.COM"
```
of course, if you define `httpsRedirect`, it will return `HTTP/1.1 301 Moved Permanently`, and your AWS LB cannot forward to the service.

**My testing environment:**
**Gateway**:
```
apiVersion: networking.istio.io/v1beta1
kind: Gateway
metadata:
  name: server-3-zufar-io-gateway
  namespace: istio-system
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - "server-3.zufar.io"
    port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
  - hosts:
    - "server-3.zufar.io"
    port:
      name: https-default
      number: 443
      protocol: HTTPS
    tls:
      credentialName: server-3-zufar-io-cert
      mode: SIMPLE
```
**VirtualService**
```
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: server-3-helloworld-01-vs
  namespace: istio-testing
spec:
  hosts:
  - "server-3.zufar.io"
  gateways:
  - istio-system/server-3-zufar-io-gateway
  http:
  - name: "helloworld-01-istio-testing-route"
    match:
    - uri:
        prefix: /web-01/
    rewrite:
      uri: /
    route:
      - destination:
          port:
            number: 80
          host: helloworld.istio-testing.svc.cluster.local
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: server-3-helloworld-02-vs
  namespace: istio-testing
spec:
  hosts:
  - "server-3.zufar.io"
  gateways:
  - istio-system/server-3-zufar-io-gateway
  http:
  - name: "helloworld-02-istio-testing-route"
    match:
    - uri:
        prefix: /web-02/
    rewrite:
      uri: /
    route:
      - destination:
          port:
            number: 80
          host: helloworld.istio-testing.svc.cluster.local
```
**curl testing**:
```
gateway $ curl http://server-3.zufar.io/web-01/ -v
*   Trying 10.200.100.110...
* TCP_NODELAY set
* Connected to server-3.zufar.io (10.200.100.110) port 80 (#0)
> GET /web-01/ HTTP/1.1
> Host: server-3.zufar.io
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< location: https://server-3.zufar.io/web-01/
< date: Thu, 05 Nov 2020 14:14:58 GMT
< server: istio-envoy
< content-length: 0
< 
* Connection #0 to host server-3.zufar.io left intact

gateway $ curl http://server-3.zufar.io/web-02/ -v
*   Trying 10.200.100.110...
* TCP_NODELAY set
* Connected to server-3.zufar.io (10.200.100.110) port 80 (#0)
> GET /web-02/ HTTP/1.1
> Host: server-3.zufar.io
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
< location: https://server-3.zufar.io/web-02/
< date: Thu, 05 Nov 2020 14:15:01 GMT
< server: istio-envoy
< content-length: 0
< 
* Connection #0 to host server-3.zufar.io left intact

gateway $ curl https://server-3.zufar.io/web-01/       
{
  "path": "/",
  "headers": {
    "host": "server-3.zufar.io",
    "user-agent": "curl/7.58.0",
    "accept": "*/*",
    "x-forwarded-for": "10.244.4.1",
    "x-forwarded-proto": "https",
    "x-request-id": "d355feec-59da-4383-a8b3-71035329cb35",
    "path-without-params": "/web-01/",
    "x-envoy-attempt-count": "1",
    "x-envoy-original-path": "/web-01/",
....
}

gateway $ curl https://server-3.zufar.io/web-02/   
{
  "path": "/",
  "headers": {
    "host": "server-3.zufar.io",
    "user-agent": "curl/7.58.0",
    "accept": "*/*",
    "x-forwarded-for": "10.244.4.1",
    "x-forwarded-proto": "https",
    "x-request-id": "412f2a60-c7dc-9733-9e47-b7ca47cd7f20",
    "path-without-params": "/web-02/",
    "x-envoy-attempt-count": "1",
    "x-envoy-original-path": "/web-02/",
...
}
```

**Comment by andrascz** [comment]
@zufardhiyaulhaq I can only repeat my former comment:
> We would like to have both HTTP routing and HTTP to HTTPS redirection on the same gateway port with HTTP protocol. The redirection happens when the `X-Forwarded-Proto` header contains `HTTP`, the routing happens when it contains `HTTPS`.

Port numbers does not matter, what matters is the protocol of the port which needs to be HTTP as the AWS LB terminates TLS.
This was a working setup with istio < 1.7 and I do not see why it should be invalid.

**Comment by zufardhiyaulhaq** [comment]
@andrascz pardon me, forget about AWS LB TLS termination behaviour, currently using NLB and TLS termination happen in Gateway object. CCMIW, the AWS LB will terminate TLS but still using the same port for forwarding to Istio right?

so for example, `user port:443 -> AWS LB downstream port:443 -> AWS LB TLS termination -> AWS LB upstream port:443 -> Istio port:443`

**Comment by andrascz** [comment]
Yes, the istio-ingressgateway Service maps the Istio Ingressgateway ports 80 and 443 to the Gateway port 80 with protocol HTTP. The problem is that in istio 1.7.x `port.protocol: HTTP` conflicts with `tls.httpsRedirect: true` which results in virtualservices not getting routed.

**Comment by zufardhiyaulhaq** [comment]
@andrascz can you just make the `Gateway` object like this one? 
```
spec:
  selector:
    istio: ingressgateway
  servers:
  - hosts:
    - 'domain.io'
    port:
      name: http
      number: 443
      protocol: HTTP
```

since I believe AWS LB is already using HTTP rather than HTTPS?

**Comment by andrascz** [comment]
We *had* to make the gateway like that to be able to upgrade to 1.7.x. Although we are using a different port number, but the port number is indifferent from the functionality point of view. Putting something with protocol HTTP on the port 443 does not make a difference versus having it on port 80 or 8888 or whatever.
This meant we had to drop the `httpsRedirect: true` feature, which is why this issue is born because we want that back.

As I stated it multiple times the problem is that setting  `port.protocol: HTTP` and `tls.httpsRedirect: true` together is deemed an invalid configuration by istio from 1.7. In case of a downstream proxy terminating TLS and providing the `X-Forwarded-Proto` header signaling the original protocol, it is a valid one in my opinion.

**Comment by SeanWallace** [comment]
Indeed the assumption made in #24958 is false:

> we should not need the http route since all http is redirected to tls

This is literally the best example why it's better to require the specification than to assume all traffic as otherwise.

**Comment by howardjohn** [comment]
Sorry for the delay. I don't have an AWS cluster to do a full reproduction, but i wasn't able to reproduce this.

I installed Istio 1.6, set up my ports to match yours (80->80, 443->80) and created this config:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: echo
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "*"
    tls:
      httpsRedirect: true
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: echo
spec:
  hosts:
  - "*"
  gateways:
  - echo
  http:
  - route:
    - destination:
        host: echo
        port:
          number: 80
```

then I sent requests:
```
curl http://ingress:443 -v -H "X-Forwarded-Proto: https"
curl http://ingress:443 -v
curl http://ingress:80 -v -H "X-Forwarded-Proto: https"
curl http://ingress:80 -v
```

All 4 of these return the same 301 response, which is what I would expect.

This behavior is the same on 1.8 and 1.6 for me. To me, this seems expected. What I am not sure is why it was working for you in 1.6.

My understanding is the ingress is going to get an HTTP request on port 8080 with X-Forwarded-Proto=https, and you expect this to NOT be redirect. However, I don't think this is the case unless you set use_remote_address=false or XffNumTrustedHops>1, as envoy will not respect the header. Those are the envoy configs, they can be configured by https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/ in Istio. See also https://github.com/envoyproxy/envoy/issues/4496.

So overall, this seems to be working as expected, but I don't understand how it worked for you previously, so likely I am missing something...

**Comment by andrascz** [comment]
Thank you for the response. You are raising an important point about XffNumTrustedHops. Sorry that I forgo tabout that part and did not include it in this issue. Here is the EnvoyFilter we are using:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: ingressgateway-redirect-config
  namespace: istio-system
spec:
  configPatches:
  - applyTo: NETWORK_FILTER
    match:
      context: GATEWAY
      listener:
        filterChain:
          filter:
            name: envoy.http_connection_manager
    patch:
      operation: MERGE
      value:
        typed_config:
          '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          xff_num_trusted_hops: 1
          normalize_path: true
  workloadSelector:
    labels:
      istio: ingressgateway

# NOTE: with this we can configure istio that we are behind a trusted load balancer
# With this the X-Forwarded-Proto is processed and trusted
# See: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#config-http-conn-man-headers-x-forwarded-for
# Based on:
#  * https://github.com/istio/istio/issues/18300#issuecomment-586595566
#  * https://github.com/istio/istio/issues/17980#issuecomment-573423797
```

**Comment by howardjohn** [comment]
@andrascz got it, thanks, that was the missing piece!

**Comment by gertvdijk** [comment]
Thanks for picking this up @howardjohn, good to see progress in the linked PR! :heart:

**Comment by gertvdijk** [comment]
Great to see this is fixed on master with 56d740a174df19204dea4962295c2e6ea44d77e3. :tada: Thank you again for working on this fix. :pray:

Any plans to backport this to 1.7.x/1.8.x as well?
